controllers/network/service_controller.go

```diff
@@ -0,0 +1,198 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// ServiceReconciler reconciles a Service object
+type ServiceReconciler struct {
+	client.Client
+	Kclient kubernetes.Interface
+	Log     logr.Logger
+	Scheme  *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=network.openstack.org,resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=network.openstack.org,resources=services/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=network.openstack.org,resources=services/finalizers,verbs=update
+// +kubebuilder:rbac:groups=network.openstack.org,resources=dnsmasqs,verbs=get;list;watch;
+// +kubebuilder:rbac:groups=network.openstack.org,resources=dnsdatas,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Service object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *ServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	// Get a list of all DNSMasqs to trigger reconcile only for these namespaces
+	dnsmasqs := &networkv1.DNSMasqList{}
+
+	listOpts := []client.ListOption{
+		client.InNamespace(req.Namespace),
+	}
+	if err := r.Client.List(ctx, dnsmasqs, listOpts...); err != nil {
+		return ctrl.Result{}, fmt.Errorf("Unable to retrieve DNSMasqList %w", err)
+	}
+
+	for _, dnsmasq := range dnsmasqs.Items {
+		dnsHosts, err := r.getServiceDNSData(ctx, dnsmasq.Namespace)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// sort entries for DNSData spec to reduce not required updates
+		sortedDNSHosts := []networkv1.DNSHost{}
+
+		keys := maps.Keys(dnsHosts)
+		sort.Strings(keys)
+
+		for _, key := range keys {
+			sortedDNSHosts = append(sortedDNSHosts, dnsHosts[key])
+		}
+
+		err = r.createOrPatchDNSData(ctx, dnsmasq.Namespace, &dnsmasq.ObjectMeta, dnsmasq.Spec.DNSDataLabelSelectorValue, sortedDNSHosts)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Service{}).
+		Complete(r)
+}
+
+// getServiceDNSData -
+func (r *ServiceReconciler) getServiceDNSData(
+	ctx context.Context,
+	namespace string,
+) (map[string]networkv1.DNSHost, error) {
+	svcDNSHosts := map[string]networkv1.DNSHost{}
+
+	// get all services from the namespace triggered the reconcile
```

Comment:
I guess we do this so that we can regenerate the whole DNSHost list for a namespace for each reconcile instead of trying to modify some specific items in DNSHost list based on the Service that is changed.  This is clearly a tradeoff between simpler code vs more efficient but more complex code. If this become a performance bottleneck then we can avoid listing all the Services in the namespace and just modify the DNSHost data for the given Service that is being reconciled here.

Reply:
correct, I had the same thinking, but since we don't have that many services in a namespace , I think its ok to get the list. When we move the operators to its own namespace we should only have the ones from the openstack services there.  what we could do is in addition the the annotation, we also add a dnsmasq label, so that we could limit the resulting list to only those.